File: solutions/1658.js
/**
 * @param {number[]} nums
 * @param {number} x
 * @return {number}
 * 
 * gy Jacky Li: 
 * Time complexity: O(n);
 * Space complexity: O(1);
 */

 // DFs
 // a sliding window to find the longest subarray length in which the sum === total - x;
 var minOperations = function(nums, x) {
    // // base or edge case: 
    let left = 0;
    let curSum = 0;
    let longest = Number.NEGATIVE_INFINITY;
    let total = nums.reduce((partialSum, a) => partialSum + a, 0);
    if (total - x < 0) {
        return -1;
    }

    for (let right = 0; right < nums.length; right++) {
        curSum += nums[right];
        
        while (curSum > total - x) {
            // move the left pointer until it is == or smaller
            curSum -= nums[left];
            left++;
        }

        // if the curSum is equal:
        if (curSum === total - x) {
            longest = Math.max(longest, right - left + 1);
        }
    }
    return longest === Number.NEGATIVE_INFINITY ? -1 : nums.length - longest;
};

File: solutions/1658.java
class Solution {
    public int minOperations(int[] nums, int x) {
        int n = nums.length;
        // find max window sum needed
        int targetSum = 0;
        for (int num : nums) { 
            targetSum += num;
        }
        targetSum = targetSum - x;

        if (targetSum == 0) {
            return n;
        }

        int sum = nums[0];
        int maxLength = 0;
        int j = 0;
        for (int i = 0 ; i < n ; i++) {
            while (sum < targetSum && j < n - 1) {
                j++;
                sum += nums[j];
            }
            if (sum == targetSum) {
                maxLength = Math.max(j - i + 1, maxLength);
            }
            sum -= nums[i];
        }

        return maxLength == 0 ? -1 : n - maxLength;
    }
}